File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

function getIndex(row, col) {
    return row * n + col;
}

function isMovable(dr, dc) {
    return 0 <= dr && dr < n && 0 <= dc && dc < n;
}

function solve(row, col) {
    const targetIndex = getIndex(row, col);

    if (memo.has(targetIndex)) return memo.get(targetIndex);

    memo.set(targetIndex, 1);
    for (let i = 0; i < 4; i++) {
        let dr = row + directionY[i];
        let dc = col + directionX[i];

        if (isMovable(dr, dc) && board[row][col] < board[dr][dc]) {
            memo.set(targetIndex, Math.max(memo.get(targetIndex), solve(dr, dc) + 1));
        }
    }

    return memo.get(targetIndex);
}

const n = Number(input[0]);
const memo = new Map();
const directionX = [0, 0, 1, -1];
const directionY = [1, -1, 0, 0];
let idx = 1;
let answer = 0;
const board = [];
for (let i = 0; i < n; i++) {
    board.push(input[idx++].split(" ").map(Number));
}

for (let r = 0; r < n; r++) {
    for (let c = 0; c < n; c++) {
        answer = Math.max(answer, solve(r, c));
    }
}

console.log(answer);

File: 백준/Gold/1937. 욕심쟁이 판다/욕심쟁이 판다.py
import sys
sys.setrecursionlimit(10**8)
input = sys.stdin.readline

def is_movable(row, col):
    return (0 <= row < n and 0 <= col < n)

def get_idx(row, col):
    return row * n + col

def solve(row, col):
    idx = get_idx(row, col)
    if (idx in memo):
        return memo[idx]
    
    memo[idx] = 1 
    for i in range(4):
        dr = row + direction_y[i]
        dc = col + direction_x[i]
        if (is_movable(dr, dc) and board[row][col] < board[dr][dc]):
            memo[idx] = max(memo[idx], solve(dr, dc) + 1)
    
    return memo[idx]

direction_x = [-1, 1, 0 ,0]
direction_y = [0, 0, 1, -1]
memo = {}
n = int(input())
board = []
for _ in range(n):
    board.append(list(map(int, input().split())))

answer = 0
for r in range(n):
    for c in range(n):
        answer = max(answer, solve(r, c))

print(answer)
